### PR TITLE
fix: disable wheel content validation

### PR DESCRIPTION
**Pull request by ralbertazzi**
# Pull Request Check List

Closes: https://github.com/python-poetry/poetry/issues/7983

First commit is a temporary solution to https://github.com/python-poetry/poetry/issues/7983 until https://github.com/pypa/installer/pull/183 gets released.

~Second commit is a permanent solution that also provides faster installation times.~

### CODE REVIEW

**Comment by ralbertazzi** [comment]
I added another commit where I achieve what I was saying here: https://github.com/python-poetry/poetry/issues/7983#issuecomment-1558645745

This:
- Requires no release of pypa/installer
- Makes the installation process faster, as we compute file hashes just once
- Does not lose any validation capabilities. They are just handled a bit later

Cons: some pypa/installer code spilling into the Poetry codebase

**Comment by radoering** [comment]
Not sure if the second commit is worth it or if we should just turn off validation until there is a solution in `installer` (or we really need validation for some reason). For a patch release I'd probably go with the first commit only (and include the second commit in the next minor if we want to include it).

> First commit is a temporary solution to https://github.com/python-poetry/poetry/issues/7983 until https://github.com/pypa/installer/pull/183 gets released.

The installer PR only reduces the memory consumption, the hash will still be computed twice, won't it?

File: src/poetry/installation/wheel_installer.py [hunk 1]
```diff
@@ -66,12 +87,52 @@ def for_source(self, source: WheelFile) -> WheelDestination:
...
+        }
+        for item in self._source._zipfile.infolist():
+            record_args = record_mapping.pop(item.filename, None)
+            if not record_args:
+                continue
+
+            file_record = RecordEntry.from_elements(*record_args)
+            computed_record = next(
+                record for _, record in records if record.path == item.filename
+            )
```
Comment (radoering) [inline src/poetry/installation/wheel_installer.py:112]: This doesn't always work. I tested a large project and got a `StopIteration` for some packages.

You can reproduce it by running `poetry add fonttools` for example.

What's even worth: The package is installed, but the `RECORD` file is not written.

`poetry run pip uninstall fonttools` now throws `ERROR: Cannot uninstall fonttools 4.39.4, RECORD file not found. Hint: The package was installed by Poetry 1.6.0.dev0.` and `poetry run pip install fonttools` gives `Requirement already satisfied`.

You can only "manually" uninstall the package.

File: src/poetry/installation/wheel_installer.py [hunk 2]
```diff
@@ -66,12 +87,52 @@ def for_source(self, source: WheelFile) -> WheelDestination:
...
+                )
+
+    def finalize_installation(
+        self,
+        scheme: Scheme,
+        record_file_path: str,
+        records: Iterable[tuple[Scheme, RecordEntry]],
+    ) -> None:
+        self._validate_hash_and_size(records)
+        return super().finalize_installation(scheme, record_file_path, records)
```
Comment (radoering) [inline src/poetry/installation/wheel_installer.py:134]:
```suggestion
        try:
            self._validate_hash_and_size(records)
        finally:
            return super().finalize_installation(scheme, record_file_path, records)
```

I wonder if we should do something like that to make sure the `RECORD` file is written despite of bugs in `_validate_hash_and_size()`. However, when I tried this, the exception was swallowed, which not good either. Thus, we would have to put more effort here...

**Review by radoering — CHANGES_REQUESTED** [review]

**Comment by ralbertazzi** [comment]
@radoering yes, I agree the PR still needs some touches. The way I got it is that we wouldn't have merged the additional work on top of the simple `validate_contents=False`, am I right?

**Comment by radoering** [comment]
Yes, I think for 1.5.1 the first commit is fine. The second commit needs more work and testing. You can split the second commit into a separate PR if you want to keep working on it.

**Review by radoering — APPROVED** [review]

**Pull request merged**